atsign-foundation/at_server
# fix: Introduce time duration for apkam keys to auto expire

**- What I did**
- Introduce APKAM keys to auto expire after specified time duration. 

**- How I did it**
- In at_commons, introduced `Duration apkamKeysExpiryDuration` to capture the time duration set by the user.  This time duration is propagated to the secondary server.
- In the secondary server, once enrollment is approved, set the TTL value to the `Duration apkamKeysExpiryDuration` in milliseconds. This will ensure that the key expires and is subsequently deleted from the keystore.
- In "AbstractVerbHandler.dart", add "_verifyIfEnrollmentIsActive", which will check if the enrollment is active on the existing connections. If the enrollment expires, closes the connection with error message.
- For the new connections, in the "PkamVerbHandler.dart->handleApkamVerification", there is call to "getEnrollDataStoreValue" which will check if the enrollment is active. If not active, sets enrollment state to expired and authentication fails.
- Introduce EnrollmentManager class which contains "get", "put" and "buildEnrollmentKey" methods which will return the enrollment details from the keystore, create/update enrollment details to the keystore and build the enrollment key for the enrollment id respectively.
- Refactor the code in the verb handler which related to fetch / creating enrollment records with happen via the enrollment manager.
- Update the unit tests to align with the EnrollmentManager class.

**- How to verify it**
- Added a unit test to verify the changes.
  - A test to verify apkam expiry is set for approved enrollment
  - A test to verify pkam verb fails when apkam keys are expired
  - A test to verify update verb fails when apkam keys are expired
  - A test to verify delete verb fails when apkam keys are expired

- Added below functional tests:
  -  A test to verify apkam authentication fails with expired apkam keys
  - A test to verify connection closes when apkam keys expire

**- Description for the changelog**
- fix: Introduce time duration for apkam keys to auto expire
